### ISSUE TITLE
to_nice_yaml producing "..." (3 dots) on role vars but not template vars

### ISSUE THREAD

**Issue by joaocc** [ansible/ansible#17095]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

template and to_nice_yaml filter
##### ANSIBLE VERSION

ansible 2.1.1.0
##### CONFIGURATION

ANSIBLE_JINJA2_EXTENSIONS=jinja2.ext.do,jinja2.ext.i18n,jinja2.ext.loopcontrols,jinja2.ext.with_
ANSIBLE_PRIVATE_ROLE_VARS=true
##### OS / ENVIRONMENT

N/A
##### SUMMARY

While attempting to use 'template' to generate a YAML file, I noticed that 'to_nice_yaml' produces trailing "..." (3 dots) after outputting some variables, but not other.
This happens with role variables that are used directly inside the template, but not with objects that are obtained through loop variables.
##### STEPS TO REPRODUCE

templates/_debug.yml.j2

```

---
{% set ii=(item[1].j2_item) %}
{% set aa=(item[1].j2_args) %}
REPRO__TO_NICE_YAML:
    fromitem:
{{ aa.rolevars.id |to_nice_yaml |indent(8,true) }}
    fromrole:
        test1:
{{ myrole__id      |to_nice_yaml |indent(10, true) }}
        test2:
{{ myrole__id      |to_nice_yaml(false,false) |indent(10, true) }}
        test3:
{{ myrole__id      |to_nice_yaml(explicit_start=False, explicit_end=False) |indent(10, true) }}

```

output file

```
```

loop_config (outer)

```
  vars:
    j2_args:
      role_vars:
        id:      "{{ myrole__id }}"
  loop_control:
    loop_var: j2_item
  with_flattened:
    - "{{ files_1 }}"
    - "{{ files_2 }}"
```

loop_config (inner)

```
- template:
    src:  "{{ item[0].path }}"
    dest: "{{ j2_item.dst }}"
  with_nested:
    - "{{ _X_j2_find_files.files }}"
    - [ {
            j2_item: "{{ j2_item }}",
            j2_args: "{{ j2_args }}"
        }
      ]
```
##### EXPECTED RESULTS

desired output file

```

---
REPRO__TO_NICE_YAML:
    fromitem:
        0base
    fromrole:
        test1:
          0base
        test2:
          0base
        test3:
          0base
```
##### ACTUAL RESULTS

```

---
REPRO__TO_NICE_YAML:
    fromitem:
        0base
        ...
    fromrole:
        test1:
          0base
          ...
        test2:
          0base
          ...
        test3:
          0base
          ...
```

**Comment by Phrozyn** [comment]
can confirm this also affects 2.2.2.0

**Comment by sivel** [comment]
I've done a little bit of research here, and the first thing that I notice, is that you seem to be passing just a plain string through `to_nice_yaml`, such as the string `0base`.  I'm not exactly sure why you would need to do that.

Based on that, and what I can tell of the yaml emitter, it looks to see if you are writing a plain string, and that plain string is the root document, and if so, it appends a line reading `...`

The `...` line indicates "end of document" in yaml (http://yaml.org/spec/1.1/#id857577)

If I use any other data type, other than a scalar, pyyaml does not insert the `...`

Based on this, I'd have to assume that you shouldn't be piping your scalar values through `to_nice_yaml` as there doesn't seem to be any need for it.

**Comment by joaocc** [comment]
Hi @sivel.
At this distance in time, I think the reason for using to_nice_yaml with a string had to do with wanting to ensure a properly formatted string (with escaping if needed), and also for flexibility in allowing the variables to contain other non-string values if needed.
While I understand your comment of "..." at end of document, shouldn't the ```explicit_end=False``` flag prevent the ```...``` from being shown?
Thanks

**Comment by sivel** [comment]
From what I could tell looking through the pyyaml code, explicit_end has no bearing on this situation.

I think at this point, this bug has less to do with ansible and more with the behavior of pyyaml.

**Comment by Phrozyn** [comment]
I had this problem yesterday, and responded to this thread, but ended up removing my comment.
It's not an ansible issue to be clear.

For me it was a journey - through jinja2 and templates.
I got the ... on a new line, but that appeared to be due to using a list of items with only one item in it (I had started with more than 1 item but ansible wants to put [] around them, and I didn't need those.
So I consolidated the list into a comma separated value with single quotes and that fixed the ... and [] issue.
